Don't set `Attribute::StackProtectStrong` when not using `CodeModel::Large`

ararslan opened the pull request:
We're currently setting `Attribute::StackProtectStrong` for all debug builds, but that can run into issues like what was observed on FreeBSD AArch64 in #59303, where the stack canary gets relocated out of the range permitted by the (there, small) code model. To avoid this problem more generally, we can just not set the strong stack protection attribute for debug builds on platforms not using a large code model. An alternative would be to use the large code model on more platforms, but some combinations of architecture and code model are not well-maintained in LLVM, so this seems safer.

Fixes #59303

gbaraldi commented:
I'm not sure how well maintained this code mode is but if you don't hit any issues then it's probably fine

ararslan commented:
How do you mean? `CodeModel::Large` is the default on x86-64. Do you mean the specific combination of AArch64 and `CodeModel::Large`? And does "maintained" here imply by us or in LLVM?

Also, if there's a concern over this specific combination of platform and code model, it's worth noting that it's only the debug build that fails, so we could perhaps condition this on `JL_DEBUG_BUILD` or whatever.

ararslan committed:
Avoid `Attribute::StackProtectStrong` for AArch64/RISC-V debug builds

When using a code model other than large, e.g. on AArch64 (small) or
RISC-V (medium), the stack canary can end up relocated outside of the
range permissible by the code model, as observed in issue 59303. To fix
this, we could either use a large code model everywhere or disable
strong stack protection for debug builds on AArch64 and RISC-V. The
former is not ideal as some combinations of platform and code model are
not as well-maintained upstream in LLVM, so let's do the latter.

Fixes issue 59303

Co-Authored-By: Sam Schweigel <sam.schweigel@juliahub.com>

ararslan commented on src/codegen.cpp:
```diff
@@ -8722,7 +8722,7 @@ static jl_llvm_functions_t
         FnAttrs.addAttribute(Attribute::NoInline);
 
-#ifdef JL_DEBUG_BUILD
+#if defined(JL_DEBUG_BUILD) && !(defined(_CPU_AARCH64_) || defined(_CPU_RISCV_))
```
Good point, comment added. Thanks!

IanButterworth commented:
Removing merge me given https://github.com/JuliaLang/julia/pull/60434#issuecomment-3690461194
